### PR TITLE
[Doc] Add pure python mode to parallelism.rst

### PR DESCRIPTION
**Pull request by matusvalo**
This PR adds pure python mode to parallelism.rst. Several minor improvements were added:

* Examples were fixed, ensured that example can be compiled and multiple fixes were introduced,
* Example was moved to a file to be checked by CI,
* Minor improvements to text formatting.

Related: #4187

### CODE REVIEW

File: docs/examples/userguide/parallelism/cimport_openmp.py [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# tag: openmp
+
+from cython.parallel import parallel
+from cython.cimports.openmp import omp_set_dynamic
+
+num_threads = cython.declare(cython.int)
+
+omp_set_dynamic(1)
+with cython.nogil, parallel():
+    num_threads = openmp.omp_get_num_threads()
```
Comment (da-woods) [inline docs/examples/userguide/parallelism/cimport_openmp.py:10]: Presumably this needs to be part of ` from cython.cimports.openmp import omp_set_dynamic`

Comment (matusvalo) [reply docs/examples/userguide/parallelism/cimport_openmp.py:10]: Hmm all examples I did check via compiling. Maybe I did copy-past error. Let me check

Comment (matusvalo) [reply docs/examples/userguide/parallelism/cimport_openmp.py]: Issue fixed. File now can be compiled by cython.

**Comment by matusvalo** [comment]
CI tests are failing converting to draft until CI is not green.

**Comment by matusvalo** [comment]
@scoder @da-woods could you give me a helping hand? I tried to move following example from the docs to file to be checked by CI:
```cython
from cython.parallel import prange

def func(double[:] x, double alpha):
    cdef Py_ssize_t i

    for i in prange(x.shape[0]):
        x[i] = alpha * x[i]
```
But compiling fails with following error:
```
Error compiling Cython file:
------------------------------------------------------------
...
from cython.parallel import prange

def func(double[:] x, double alpha):
    cdef Py_ssize_t i

    for i in prange(x.shape[0]):
                   ^
------------------------------------------------------------

memoryview_sum.pyx:6:19: prange() can only be used without the GIL

When I tried to mark the function as `nogil` the compilation is successfull:
```cython
from cython.parallel import prange

cdef void func(double[:] x, double alpha) nogil:
    cdef Py_ssize_t i

    for i in prange(x.shape[0]):
        x[i] = alpha * x[i]
```
```
$ cythonize -3i memoryview_sum.pyx
Compiling /Users/matus/dev/cython/docs/examples/userguide/parallelism/memoryview_sum.pyx because it changed.
[1/1] Cythonizing /Users/matus/dev/cython/docs/examples/userguide/parallelism/memoryview_sum.pyx
warning: memoryview_sum.pyx:7:24: Use boundscheck(False) for faster access
warning: memoryview_sum.pyx:7:9: Use boundscheck(False) for faster access
running build_ext
building 'memoryview_sum' extension
creating /Users/matus/dev/cython/docs/examples/userguide/parallelism/tmpib_kb51k/Users
creating /Users/matus/dev/cython/docs/examples/userguide/parallelism/tmpib_kb51k/Users/matus
creating /Users/matus/dev/cython/docs/examples/userguide/parallelism/tmpib_kb51k/Users/matus/dev
creating /Users/matus/dev/cython/docs/examples/userguide/parallelism/tmpib_kb51k/Users/matus/dev/cython
creating /Users/matus/dev/cython/docs/examples/userguide/parallelism/tmpib_kb51k/Users/matus/dev/cython/docs
creating /Users/matus/dev/cython/docs/examples/userguide/parallelism/tmpib_kb51k/Users/matus/dev/cython/docs/examples
creating /Users/matus/dev/cython/docs/examples/userguide/parallelism/tmpib_kb51k/Users/matus/dev/cython/docs/examples/userguide
creating /Users/matus/dev/cython/docs/examples/userguide/parallelism/tmpib_kb51k/Users/matus/dev/cython/docs/examples/userguide/parallelism
clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk -I/usr/local/include -I/usr/local/opt/openssl@1.1/include -I/usr/local/opt/sqlite/include -I/Users/matus/dev/cython39/include -I/usr/local/opt/python@3.9/Frameworks/Python.framework/Versions/3.9/include/python3.9 -c /Users/matus/dev/cython/docs/examples/userguide/parallelism/memoryview_sum.c -o /Users/matus/dev/cython/docs/examples/userguide/parallelism/tmpib_kb51k/Users/matus/dev/cython/docs/examples/userguide/parallelism/memoryview_sum.o
/Users/matus/dev/cython/docs/examples/userguide/parallelism/memoryview_sum.c:5426:28: warning: equality comparison with extraneous parentheses [-Wparentheses-equality]
  if (((__pyx_v_c_mode[0]) == 'f')) {
       ~~~~~~~~~~~~~~~~~~~~^~~~~~
/Users/matus/dev/cython/docs/examples/userguide/parallelism/memoryview_sum.c:5426:28: note: remove extraneous parentheses around the comparison to silence this warning
  if (((__pyx_v_c_mode[0]) == 'f')) {
      ~                    ^     ~
/Users/matus/dev/cython/docs/examples/userguide/parallelism/memoryview_sum.c:5426:28: note: use '=' to turn this equality comparison into an assignment
  if (((__pyx_v_c_mode[0]) == 'f')) {
                           ^~
                           =
/Users/matus/dev/cython/docs/examples/userguide/parallelism/memoryview_sum.c:3321:13: warning: unused function '__pyx_f_14memoryview_sum_func' [-Wunused-function]
static void __pyx_f_14memoryview_sum_func(__Pyx_memviewslice __pyx_v_x, double __pyx_v_alpha) {
            ^
2 warnings generated.
clang -bundle -undefined dynamic_lookup -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk /Users/matus/dev/cython/docs/examples/userguide/parallelism/tmpib_kb51k/Users/matus/dev/cython/docs/examples/userguide/parallelism/memoryview_sum.o -L/usr/local/lib -L/usr/local/opt/openssl@1.1/lib -L/usr/local/opt/sqlite/lib -o /Users/matus/dev/cython/docs/examples/userguide/parallelism/memoryview_sum.cpython-39-darwin.so
```

But CI is failing in this example because CI defines `CYTHON_REFNANNY` - here is the output when I manually compile .c file:
```
$ clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -DCYTHON_REFNANNY=1 -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk -I/usr/local/include -I/usr/local/opt/openssl@1.1/include -I/usr/local/opt/sqlite/include -I/Users/matus/dev/cython39/include -I/usr/local/opt/python@3.9/Frameworks/Python.framework/Versions/3.9/include/python3.9 -c memoryview_sum.c -o memoryview_sum.o
memoryview_sum.c:3403:29: error: use of undeclared identifier '__pyx_refnanny'; did you mean '__Pyx_RefNanny'?
                            __Pyx_GOTREF(__pyx_parallel_exc_type);
                            ^
memoryview_sum.c:1582:51: note: expanded from macro '__Pyx_GOTREF'
  #define __Pyx_GOTREF(r)  __Pyx_RefNanny->GOTREF(__pyx_refnanny, (PyObject *)(r), (__LINE__))
                                                  ^
memoryview_sum.c:1551:35: note: '__Pyx_RefNanny' declared here
  static __Pyx_RefNannyAPIStruct *__Pyx_RefNanny = NULL;
                                  ^
memoryview_sum.c:3453:15: error: use of undeclared identifier '__pyx_refnanny'; did you mean '__Pyx_RefNanny'?
              __Pyx_GIVEREF(__pyx_parallel_exc_type);
              ^
memoryview_sum.c:1583:52: note: expanded from macro '__Pyx_GIVEREF'
  #define __Pyx_GIVEREF(r) __Pyx_RefNanny->GIVEREF(__pyx_refnanny, (PyObject *)(r), (__LINE__))
                                                   ^
memoryview_sum.c:1551:35: note: '__Pyx_RefNanny' declared here
  static __Pyx_RefNannyAPIStruct *__Pyx_RefNanny = NULL;
                                  ^
memoryview_sum.c:5414:28: warning: equality comparison with extraneous parentheses [-Wparentheses-equality]
  if (((__pyx_v_c_mode[0]) == 'f')) {
       ~~~~~~~~~~~~~~~~~~~~^~~~~~
memoryview_sum.c:5414:28: note: remove extraneous parentheses around the comparison to silence this warning
  if (((__pyx_v_c_mode[0]) == 'f')) {
      ~                    ^     ~
memoryview_sum.c:5414:28: note: use '=' to turn this equality comparison into an assignment
  if (((__pyx_v_c_mode[0]) == 'f')) {
                           ^~
                           =
1 warning and 2 errors generated.
```

Am I doing something wrong or there is a bug in cython?

**Comment by da-woods** [comment]
The refnanny thing definitely seems like a bug - it's probably a mistake where something doesn't bring in the right utility code.

I suspect you don't want to mark the whole function as nogil. Instead you just want to release the GIL for the `prange` similar to the example above it

```
    for i in prange(x.shape[0], nogil=True):
```

**Comment by matusvalo** [comment]
Putting `nogil=True`  fixed the issue. PR ready for review.

File: docs/examples/userguide/parallelism/breaking_loop.py [hunk 2]
```diff
@@ -0,0 +1,15 @@
+from cython.parallel import prange
```
Comment (da-woods) [inline docs/examples/userguide/parallelism/breaking_loop.py:1]: I think you've forgotten to check breaking_loop.pyx into git

File: docs/examples/userguide/parallelism/memoryview_sum.pyx [hunk 3]
```diff
@@ -0,0 +1,9 @@
+from cython.parallel import prange
+
+
+
+def func(double[:] x, double alpha):
```
Comment (da-woods) [inline docs/examples/userguide/parallelism/memoryview_sum.pyx:5]: This doesn't quite look like it'll line up with the .py file

**Review by da-woods — COMMENTED** [review]
A couple of small comments.

Most of this looks uncontroversial though.

File: docs/examples/userguide/parallelism/breaking_loop.py (hunk 2 above)
Comment (matusvalo) [reply docs/examples/userguide/parallelism/breaking_loop.py:1]: breaking_loop.pyx is already in GIT https://github.com/cython/cython/blob/master/docs/examples/userguide/parallelism/breaking_loop.pyx, there are no changes needed there so it is not included in PR.

Comment (da-woods) [reply docs/examples/userguide/parallelism/breaking_loop.py:1]: Sorry you're quite right. Clearly I wasn't being very clever earlier...

Comment (matusvalo) [reply docs/examples/userguide/parallelism/breaking_loop.py:1]: But now I see that at least .pyx file needs alignment. I will fix it now.

Comment (matusvalo) [reply docs/examples/userguide/parallelism/breaking_loop.py:1]: alignment fixed.

File: docs/examples/userguide/parallelism/memoryview_sum.pyx (hunk 3 above)
Comment (matusvalo) [reply docs/examples/userguide/parallelism/memoryview_sum.pyx:3]: fixed.

**Comment by matusvalo** [comment]
I have fixed all comments. Please do another round of review.

**Comment by da-woods** [comment]
This looks good to me. I'm going to leave it another couple of days just in case anyone else has any comments. But I'll merge it shortly otherwise

**Review by da-woods — APPROVED** [review]

**Comment by da-woods** [comment]
Thanks

**Pull request merged**